api/transport/header.go

```diff
@@ -65,8 +67,12 @@ func (h Headers) With(k, v string) Headers {
 	if h.items == nil {
 		h.items = make(map[string]string)
 	}
+	if h.rawItems == nil {
+		h.rawItems = make(map[string]string)
+	}
 
 	h.items[CanonicalizeHeaderKey(k)] = v
+	h.rawItems[k] = v
```

Comment:
We previously had a guarantee that the last equivalent header would win for the purposes of equivalent header collisions. We need to preserve that invariant.

To do that, you need a single map from canonical key to a struct of { raw key and value }. That way you can easily overwrite the item if the canonical key matches an existing canonical key.